src/java/org/apache/cassandra/index/sai/disk/v1/SSTableIndexWriter.java

```diff
@@ -0,0 +1,373 @@
+
+import com.google.common.base.Stopwatch;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.db.marshal.AbstractType;
+import org.apache.cassandra.db.rows.Row;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.analyzer.AbstractAnalyzer;
+import org.apache.cassandra.index.sai.disk.PerIndexWriter;
+import org.apache.cassandra.index.sai.disk.format.IndexComponent;
+import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
+import org.apache.cassandra.index.sai.utils.NamedMemoryLimiter;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.TypeUtil;
+import org.apache.cassandra.utils.Clock;
+import org.apache.cassandra.utils.FBUtilities;
+import org.apache.cassandra.utils.NoSpamLogger;
+
+/**
+ * Column index writer that accumulates (on-heap) indexed data from a compacted SSTable as it's being flushed to disk.
+ */
+@NotThreadSafe
+public class SSTableIndexWriter implements PerIndexWriter
+{
+    private static final Logger logger = LoggerFactory.getLogger(SSTableIndexWriter.class);
+    private static final NoSpamLogger noSpamLogger = NoSpamLogger.getLogger(logger, 1, TimeUnit.MINUTES);
+
+    public static final int MAX_STRING_TERM_SIZE = Integer.getInteger("cassandra.sai.max_string_term_size_kb", 1) * 1024;
+    public static final int MAX_FROZEN_TERM_SIZE = Integer.getInteger("cassandra.sai.max_frozen_term_size_kb", 5) * 1024;
+    public static final String TERM_OVERSIZE_MESSAGE = "Can't add term of column {} to index for key: {}, term size {} " +
+                                                       "max allowed size {}, use analyzed = true (if not yet set) for that column.";
+
+    private final IndexDescriptor indexDescriptor;
+    private final IndexContext indexContext;
+    private final int nowInSec = FBUtilities.nowInSeconds();
+    private final AbstractAnalyzer analyzer;
+    private final NamedMemoryLimiter limiter;
+    private final int maxTermSize;
+    private final BooleanSupplier isIndexValid;
+
+    private boolean aborted = false;
+
+    // segment writer
+    private SegmentBuilder currentBuilder;
+    private final List<SegmentMetadata> segments = new ArrayList<>();
+
+    public SSTableIndexWriter(IndexDescriptor indexDescriptor, IndexContext indexContext, NamedMemoryLimiter limiter, BooleanSupplier isIndexValid)
+    {
+        this.indexDescriptor = indexDescriptor;
+        this.indexContext = indexContext;
+        this.analyzer = indexContext.getIndexAnalyzerFactory().create();
+        this.limiter = limiter;
+        this.isIndexValid = isIndexValid;
+        this.maxTermSize = indexContext.isFrozen() ? MAX_FROZEN_TERM_SIZE : MAX_STRING_TERM_SIZE;
+    }
+
+    @Override
+    public void addRow(PrimaryKey key, Row row, long sstableRowId) throws IOException
+    {
+        if (maybeAbort())
+            return;
+
+        if (indexContext.isNonFrozenCollection())
```

Comment:
What happens w/ frozen collections? I'm having trouble finding a test for that...

Reply:
`FrozenSetAsciiTest` would exercise this if I had added back the correct code to `IndexingTypeSupport`. I've added it now so you will be able to see it used.